### PR TITLE
add flags for preEIP155 and free transactions

### PR DESCRIPTION
**Pull request by kstoykov**

### CODE REVIEW

File: cmd/hack/rpc-trace-compare/main.go [hunk 1]
```diff
@@ -59,9 +59,9 @@ func main() {
 			return
 		}
 
-		if err = os.Remove("./traces/" + traceFile); err != nil {
-			fmt.Println(err)
-		}
+		// if err = os.Remove("./traces/" + traceFile); err != nil {
```
Comment (mandrigin) [inline cmd/hack/rpc-trace-compare/main.go:62]: seems like an unnecessary change for me

File: cmd/rpcdaemon/commands/daemon.go [hunk 2]
```diff
@@ -18,10 +18,11 @@ func APIList(db kv.RoDB, borDb kv.RoDB, eth rpchelper.ApiBackend, txPool txpool.
 	filters *rpchelper.Filters, stateCache kvcache.Cache,
 	blockReader services.FullBlockReader, agg *libstate.AggregatorV3, cfg httpcfg.HttpCfg, engine consensus.EngineReader,
 	l2RpcUrl string,
+	allowFreeTransactions, allowPreEIP155Transactions bool,
 ) (list []rpc.API) {
 	base := NewBaseApi(filters, stateCache, blockReader, agg, cfg.WithDatadir, cfg.EvmCallTimeout, engine, cfg.Dirs)
 	base.SetL2RpcUrl(l2RpcUrl)
-	ethImpl := NewEthAPI(base, db, eth, txPool, mining, cfg.Gascap, cfg.ReturnDataLimit, "")
+	ethImpl := NewEthAPI(base, db, eth, txPool, mining, cfg.Gascap, cfg.ReturnDataLimit, "", allowFreeTransactions, allowPreEIP155Transactions)
```
Comment (mandrigin) [inline cmd/rpcdaemon/commands/daemon.go]: there are a few parameters that are zk-related, maybe group then into a struct? l1rpc, and these flags?

File: cmd/utils/flags.go [hunk 3]
```diff
@@ -457,6 +457,16 @@ var (
 		Usage: "Define the host used for the zkevm data stream",
 		Value: "",
 	}
+	AllowFreeTransactions = cli.UintFlag{
+		Name:  "zkevm.allow-free-transactions",
+		Usage: "Allow the sequencer to proceed transactions with 0 gas",
```
Comment (mandrigin) [inline cmd/utils/flags.go]: is it 0 gas or 0 gas price?

File: cmd/utils/flags.go [hunk 4]
```diff
@@ -457,6 +457,16 @@ var (
...
+		Value: 0,
+	}
+	AllowPreEIP155Transactions = cli.UintFlag{
+		Name:  "zkevm.allow-pre-eip155-transactions",
+		Usage: "Allow the sequencer to proceed pre-EIP155 transactions",
+		Value: 0,
```
Comment (mandrigin) [inline cmd/utils/flags.go]: there is no boolean flag?

File: cmd/utils/flags.go (hunk 3 above)
Comment (mandrigin) [reply cmd/utils/flags.go]: also, is there no boolean flag?

**Review by mandrigin — CHANGES_REQUESTED** [review]
some changes needed

**Comment by mandrigin** [comment]
@hexoscott I am pretty sure you need to review that as well

File: cmd/hack/rpc-trace-compare/main.go (hunk 1 above)
Comment (kstoykov) [reply cmd/hack/rpc-trace-compare/main.go:62]: There is no point to delete traces when comparing the traces. This is undesired side effect.

File: cmd/rpcdaemon/commands/daemon.go (hunk 2 above)
Comment (kstoykov) [reply cmd/rpcdaemon/commands/daemon.go]: All zk-related are currently in Zk struct. There are 22 flags, some of them are prefixed by "L2" or "L1" and some are non-prefixed. By saying a "struct" do you really mean to make a sub-struct inside Zk struct or you meant adding a prefix?

Comment (hexoscott) [reply cmd/rpcdaemon/commands/daemon.go]: There were changes merged yesterday which could help here.  We already pass the l2RPCUrl and now with two more ZK related things it makes sense to just pass in the full `ZK` config struct and to add these new properties to that.  If you rebase this branch onto zkevm you'll see the changes made there as they'll conflict

File: cmd/utils/flags.go (hunk 3 above)
Comment (kstoykov) [reply cmd/utils/flags.go]: both are corrected

File: cmd/utils/flags.go (hunk 4 above)
Comment (kstoykov) [reply cmd/utils/flags.go]: corrected

**Comment by hexoscott** [comment]
I thinnk it looks OK once the point around the multiple parameters being passed when we can use the ZK config struct is resolved

File: cmd/rpcdaemon/commands/daemon.go (hunk 2 above)
Comment (kstoykov) [reply cmd/rpcdaemon/commands/daemon.go]: I will update it so it will use the entire struct

Comment (kstoykov) [reply cmd/rpcdaemon/commands/daemon.go]: Done

File: cmd/rpcdaemon/commands/eth_api.go [hunk 5]
```diff
@@ -322,32 +322,36 @@ func (api *BaseAPI) getEffectiveGasPricePercentage(tx kv.Tx, txHash common.Hash)
...
+	GasCap                     uint64
+	ReturnDataLimit            int
+	ZkRpcUrl                   string
+	AllowFreeTransactions      bool
+	AllowPreEIP155Transactions bool
 }
 
 // NewEthAPI returns APIImpl instance
-func NewEthAPI(base *BaseAPI, db kv.RoDB, eth rpchelper.ApiBackend, txPool txpool.TxpoolClient, mining txpool.MiningClient, gascap uint64, returnDataLimit int, zkRpcUrl string) *APIImpl {
+func NewEthAPI(base *BaseAPI, db kv.RoDB, eth rpchelper.ApiBackend, txPool txpool.TxpoolClient, mining txpool.MiningClient, gascap uint64, returnDataLimit int, zkRpcUrl string, allowFreeTransactions, allowPreEIP155Transactions bool) *APIImpl {
```
Comment (hexoscott) [inline cmd/rpcdaemon/commands/eth_api.go]: could the last three parameters here be replaced by just an `ethconfig.ZK`?

Comment (kstoykov) [reply cmd/rpcdaemon/commands/eth_api.go]: Technically - yes but I saw that sometimes the zkRpcUrl is empty while allowFreeTransactions, allowPreEIP155Transactions are not. If we replace the three params with a single struct then we will not be able to achieve this. You can check this in AuthAPIList function (daemon.go). Tbh I have no idea why zkRpcUrl was to "" there.

Comment (hexoscott) [reply cmd/rpcdaemon/commands/eth_api.go]: Ah ok, right now we're aware that ZK will only work correctly when running erigon as a single unit.  It has the ability to break off parts, such as rpcdaemon.  Currently these won't work in a ZK context so we just pass in "" as the URL because the application just won't work like that.  Something we need to address for sure longer term, but right now you can pass dummy stuff there as it won't work anyway, the focus is on wherever we're passing a "real" value

Comment (kstoykov) [reply cmd/rpcdaemon/commands/eth_api.go]: I see. So we're going to use the 3 params rather than a struct for now, right?

Comment (hexoscott) [reply cmd/rpcdaemon/commands/eth_api.go]: For me the struct is a nicer approach.  The other places that aren't being used right now (like rpcdaemon) can just take the default to make the compiler happy, we can address all of those in one go later.

Comment (kstoykov) [reply cmd/rpcdaemon/commands/eth_api.go]: done

**Review by mandrigin — APPROVED** [review]

**Review by mandrigin — APPROVED** [review]
